### PR TITLE
feat(eips): Make prague field an enum

### PR DESCRIPTION
**Pull request by DoTheBestToGetTheBest**
closes https://github.com/alloy-rs/alloy/issues/1557

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, can we also integrate this into the ExecutionPayloadSidecar type?

File: crates/eips/src/eip7685.rs [hunk 1]
```diff
@@ -76,7 +76,29 @@ impl Requests {
         self.0.extend(other.take());
     }
 }
+/// A list of requests or a precomputed requests hash.
```
Comment (tcoratger) [inline crates/eips/src/eip7685.rs:79]:
```suggestion
}

/// A list of requests or a precomputed requests hash.
```

File: crates/eips/src/eip7685.rs [hunk 2]
```diff
@@ -76,7 +76,29 @@ impl Requests {
...
+    /// Stores a precomputed requests hash.
+    Hash(B256),
+}
 
+impl PragueRequests {
+    /// Returns the requests hash for the enum instance.
+    ///
+    /// - If the instance contains a list of requests, this function calculates the hash using
+    ///   `requests_hash` of the `Requests` struct.
+    /// - If it contains a precomputed hash, it returns that hash directly.
```
Comment (tcoratger) [inline crates/eips/src/eip7685.rs:93]:
```suggestion
    /// Returns the requests hash for the enum instance.
    ///
    /// - If the instance contains a list of requests, this function calculates the hash using
    ///   `requests_hash` of the [`Requests`] struct.
    /// - If it contains a precomputed hash, it returns that hash directly.
```

**Comment by DoTheBestToGetTheBest** [comment]
> cool, can we also integrate this into the ExecutionPayloadSidecar type?

done thanks for the review

File: crates/eips/src/eip7685.rs (hunk 2 above)
Comment (DoTheBestToGetTheBest) [reply crates/eips/src/eip7685.rs]: thanks @tcoratger

File: crates/eips/src/eip7685.rs (hunk 1 above)
Comment (DoTheBestToGetTheBest) [reply crates/eips/src/eip7685.rs]: ty!

File: crates/eips/src/eip7685.rs [hunk 3]
```diff
@@ -77,6 +77,30 @@ impl Requests {
     }
 }
 
+/// A list of requests or a precomputed requests hash.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum PragueRequests {
```
Comment (onbjerg) [inline crates/eips/src/eip7685.rs:83]: please find a better name like `RequestsOrHash`, this is very vague

File: crates/rpc-types-engine/src/sidecar.rs [hunk 4]
```diff
@@ -51,6 +51,10 @@ impl ExecutionPayloadSidecar {
 
     /// Returns the EIP-7685 requests
     pub const fn requests(&self) -> Option<&Requests> {
-        self.prague.as_ref()
+        if let Some(PragueRequests::Requests(ref requests)) = self.prague {
```
Comment (onbjerg) [inline crates/rpc-types-engine/src/sidecar.rs:54]: please update the doc comment to reflect that this might not be requests anymore

**Review by onbjerg — CHANGES_REQUESTED** [review]

File: crates/eips/src/eip7685.rs (hunk 3 above)
Comment (onbjerg) [reply crates/eips/src/eip7685.rs:83]: additionally, we should note that the hash is for testing purposes only, so please add more doc comments on why this enum exists

File: crates/rpc-types-engine/src/sidecar.rs (hunk 4 above)
Comment (onbjerg) [reply crates/rpc-types-engine/src/sidecar.rs:54]: this also isn't super helpful for the motivation of the issue; we should probably add a `requests_hash` method as well on the sidecar.

**Comment by DoTheBestToGetTheBest** [comment]
@onbjerg hey sorry for the delay i'm having internet problem at home, does the change look good to you ?

**Review by yash-atreya — APPROVED** [review]
lgtm! cc @mattsse

File: crates/eips/src/eip7685.rs [hunk 5]
```diff
@@ -77,6 +77,35 @@ impl Requests {
     }
 }
 
+/// A list of requests or a precomputed requests hash.
+///
+/// For testing purposes, the `Hash` variant stores a precomputed requests hash. This can be useful
+///   when the exact contents of the requests are unnecessary, and only a consistent hash value is
+/// needed   to simulate the presence of requests without holding actual data.
```
Comment (onbjerg) [inline crates/eips/src/eip7685.rs:84]: the whitespaces in this comment are a bit funky, please reformat it :D

File: crates/eips/src/eip7685.rs [hunk 6]
```diff
@@ -77,6 +77,35 @@ impl Requests {
...
+
```
Comment (onbjerg) [inline crates/eips/src/eip7685.rs:85]:
```suggestion
```

File: crates/rpc-types-engine/src/sidecar.rs [hunk 7]
```diff
@@ -51,6 +51,20 @@ impl ExecutionPayloadSidecar {
...
+
+    /// Calculates or retrieves the requests hash.
+    ///
+    /// - If the `prague` field contains a list of requests, it calculates the requests hash
+    ///   dynamically.
+    /// - If it contains a precomputed hash (used for testing), it returns that hash directly.
+
+    pub fn requests_hash(&self) -> B256 {
+        self.prague.as_ref().unwrap().requests_hash()
     }
```
Comment (onbjerg) [inline crates/rpc-types-engine/src/sidecar.rs:69]: Please make this an `Option<B256>`, we don't want to unwrap here. In the case `self.prague` is `None`, we should just return `None`.

File: crates/rpc-types-engine/src/sidecar.rs [hunk 8]
```diff
@@ -51,6 +51,20 @@ impl ExecutionPayloadSidecar {
...
+            None
+        }
+    }
+
+    /// Calculates or retrieves the requests hash.
+    ///
+    /// - If the `prague` field contains a list of requests, it calculates the requests hash
+    ///   dynamically.
+    /// - If it contains a precomputed hash (used for testing), it returns that hash directly.
+
```
Comment (onbjerg) [inline crates/rpc-types-engine/src/sidecar.rs:66]:
```suggestion
```

**Review by onbjerg — DISMISSED** [review]

File: crates/rpc-types-engine/src/sidecar.rs (hunk 7 above)
Comment (DoTheBestToGetTheBest) [reply crates/rpc-types-engine/src/sidecar.rs]: > Please make this an `Option<B256>`, we don't want to unwrap here. In the case `self.prague` is `None`, we should just return `None`.

done

File: crates/eips/src/eip7685.rs (hunk 5 above)
Comment (DoTheBestToGetTheBest) [reply crates/eips/src/eip7685.rs]: > the whitespaces in this comment are a bit funky, please reformat it :D

done

**Review by klkvr — APPROVED** [review]

**Pull request merged**